doc(chore) update ipv4 content and add doc links (#1861)
questdb/questdb.io: blog/2023-08-07-questdb-release-7.3-deduplication-ipv4.mdx

### Support for IPv4 

> IPv4 addresses are a common data type with their own unique numerical convention. We often hear from teams who are tracking the activity of users, servers or sensors, that managing IP addresses without validity checks and type-specific functions. 

Reviewer: ```suggestion
IPv4 addresses are a common data type with their own unique numerical convention. We often hear from teams who are tracking the activity of users, servers or sensors, that managing IP addresses without validity checks and type-specific functions can be frustrating. 
```

Those working with IPs will appreciate IPv4 support, which is more flexible and performant with SQL execution, and much kinder on storage demand. 

### Are all forms of IPv4 address supported?

> Yes! IPv4 addresses are stored as integers on disk, and passed as strings.

Reviewer: ```suggestion
Yes! IPv4 addresses are stored as integers on disk, and passed as strings. IPv4 addresses may not contain netmasks when being stored on disk, however, when they are being passed as a string argument to a function (e.g., subnet inclusion operators: <<=, <<) they may also take the form of ip address with netmask or subnet with netmask (see below).
```

IPv4 arguments may appear as: 

* an ip address without netmask: `23.62.200.1`

String arguments may appear as: 

* an ip address without netmask: `23.62.200.1`
* an ip address with netmask: `5.43.11.8/2` 
* a subnet with netmask: `2.232/16` 

Altogether, we support the full range of IPv4 addresses.

### Which SQL operations can be performed on IPv4 addresses?

The following operations are supported:

* `=`: is an ip address equal to another? (`boolean`)

> * `<<`: is an ip address strictly contained by subnet? (`boolean`)

Reviewer: ```suggestion
* `<<`: is an ip address strictly contained by a subnet? (`boolean`)
```

* `<<=`: is an ip address contained by or equal to subnet? (`boolean`)
* `&`: computes bitwise AND (`ipv4`)
* `~`: computes bitwise NOT (`ipv4`)
* `|`: computes bitwise OR (`ipv4`)
* `+`: adds an offset to an ip address (`ipv4`)
* `-`: `ipv4-int`: subtracts an offset from an ip address (`ipv4`)
* `-`: `ipv4 - ipv4`: computes the difference between two ip addresses (`long`)

> There are also two additional functions for testing:

Reviewer: ```suggestion
There are also two additional random functions for testing as well as a netmask function:
```

* `rnd_ipv4()`: random address generator (String)
* `rnd_ipv4(string, int)`: random address generator within a range (String)
* `netmask(string)` : returns an ip address' subnet mask (`ipv4`)

For a full list of operators, see the [IPv4 documentation](/docs/reference/operators/ipv4/).

### Which SQL functions can be performed on IPv4 columns?

> The IPv4 data type supports SQL functions, like `CASE`, `NULLIF`, `UNION ALL` and many more. For a full list, [checkout our documentation](/docs/reference/function/ipv4/).

Reviewer: ```suggestion
The IPv4 data type supports SQL functions, like `ORDER BY`, `SAMPLE BY`, `LATEST BY` and many more. For a full list, [checkout our documentation](/docs/reference/function/ipv4/).
```

> 

Reviewer: Might be worth adding a heading and link here redirecting users to a page on the similarities/differences between ipv4 and inet like for the SQL functions - I'll start working on the content for that table now

Author: We'll tackle in separate PR. 👍

Yes! IPv4 addresses are stored as integers on disk, and passed as strings. 

Do note that IPv4 addresses may not contain netmasks when being stored on disk. However, when they are being passed as a string argument to a function - like with subnet inclusion operators: `<<=` & `<<` - they may include netmask or subnet with a netmask as demonstrated below.

IPv4 arguments may appear as: 

* an ip address without netmask: `23.62.200.1`

String arguments may appear as: 

* an ip address without netmask: `23.62.200.1`
* an ip address with netmask: `5.43.11.8/2` 
* a subnet with netmask: `2.232/16` 

Altogether, we support the full range of IPv4 addresses.

### Which SQL operations can be performed on IPv4 addresses?

The following operations are supported:

* `=`: is an ip address equal to another? (`boolean`)
* `<<`: is an ip address strictly contained by a subnet? (`boolean`)
* `<<=`: is an ip address contained by or equal to a subnet? (`boolean`)
* `&`: computes bitwise AND (`ipv4`)
* `~`: computes bitwise NOT (`ipv4`)
* `|`: computes bitwise OR (`ipv4`)
* `+`: adds an offset to an ip address (`ipv4`)
* `-`: `ipv4-int`: subtracts an offset from an ip address (`ipv4`)

> * `-`: `ipv4 - ipv4`: computes the difference between two ip addresses (`long`)

Reviewer: Could you update the operators to reflect this pr https://github.com/questdb/questdb/pull/3650 please? Thanks!

Author: On it now, in a separate PR I am also adding them to docs.

Author: Included in our our list.

& to merge after: https://github.com/questdb/questdb.io/pull/1865

Reviewer: Great!

Yes! IPv4 addresses are stored as integers on disk, and passed as strings.

IPv4 arguments may appear as: 

* an ip address without netmask: `23.62.200.1`

String arguments may appear as: 

* an ip address without netmask: `23.62.200.1`
* an ip address with netmask: `5.43.11.8/2` 
* a subnet with netmask: `2.232/16` 

Altogether, we support the full range of IPv4 addresses.

### Which SQL operations can be performed on IPv4 addresses?

The following operations are supported:

* `=`: is an ip address equal to another? (`boolean`)
* `<<`: is an ip address strictly contained by subnet? (`boolean`)
* `<<=`: is an ip address contained by or equal to subnet? (`boolean`)
* `&`: computes bitwise AND (`ipv4`)
* `~`: computes bitwise NOT (`ipv4`)
* `|`: computes bitwise OR (`ipv4`)
* `+`: adds an offset to an ip address (`ipv4`)
* `-`: `ipv4-int`: subtracts an offset from an ip address (`ipv4`)
* `-`: `ipv4 - ipv4`: computes the difference between two ip addresses (`long`)

There are also two additional functions for testing:

* `rnd_ipv4()`: random address generator (String)
* `rnd_ipv4(string, int)`: random address generator within a range (String)

> * `netmask()` needs more 

Reviewer: ```suggestion
* `netmask(string)` : returns an ip address' subnet mask (`ipv4`)
```

For a full list of operators, see the [IPv4 documentation](/docs/reference/operators/ipv4/).

### Which SQL functions can be performed on IPv4 columns?

The following functions are supported: 

* latest by
* order by 
* join 

> * add more..

Reviewer: Supported SQL funcs: 

- case e.g., `select ip, bytes, case when ip << '2.65.32.1/2' then 1 else 0 end from test`
- count e.g., `select count(ip), bytes from test`
- first e.g., `select first(ip) from test`
- last e.g., `select last(ip) from test`
- full join e.g., `select a.count, a.ip, b.ip2, b.count2 from '*!*test' a join '*!*test2' b on b.ip2 = a.ip`
- group by e.g., `select count(count), ip from test group by ip`
- count_distinct e.g., `select count_distinct(ip) from test`
- distinct e.g., `select distinct ip from test order by ip`
- select e.g., `select ip from x except select ip2 from y`
- intersect e.g., `select ip from x intersect select ip2 from y`
- isOrdered e.g., `select isOrdered(ip) from test`
- nullif e.g., `select k, nullif(ip, '0.0.0.5') from test`
- rank e.g., `select ip, bytes, rank() over (order by ip asc) rank from test order by rank`
- union eg., `select ip from x union select ip2 from y`
- union all e.g., `select ip from x union all select ip2 from y`
- inner join e.g., `select test.count, test2.count2 from test inner join test2 on test2.ip2 = test.ip`
- latest by e.g., `select * from test latest by ip`
- left join e.g., `select test.ip, test2.ip2, test.count, test2.count2 from test left join test2 on test2.ip2 = test.ip`
- max e.g., `select max(ip) from test`
- min e.g., `select min(ip) from test`
- order by e.g., `select * from test order by ip, bytes, ts`
- sample by e.g., `select ip, ts, sum(bytes) from test sample by 1y order by 2,1`
- where e.g., `select * from test where ip = '0.0.0.1'`

Author: @allegraharris Thanks!

Release posts are more at-a-glance, so instead of lining all of this out here, created a new ipv4 page under functions. It made sense in a table, and is now part of this PR.

Reviewer: Great, thank you!

> ### QuestDB IPv4 vs PostegreSQL inet 

Author: @allegraharris Going to pull this out for now, too, and we can add it into the docs. IMO we're covered at-a-glance.

> 

Reviewer: Line 107 - we don't provide native support for subnets / netmasks 
We do provide validity checking + ipv4 specific functions which would be better to discuss here

### Are all forms of IPv4 address supported?

Yes! IPv4 addresses are stored as integers on disk, and passed as strings.

String arguments may be appear as: 

> 

Reviewer: Lines 115-134 need some reworking

Reviewer: Lines 115-119 should be cut

Reviewer: 127-134 should be changed to this: 

```sql
* `<<`: is an ip address strictly contained by subnet? (`boolean`)
* `<<=`: is an ip address contained by or equal to subnet? (`boolean`)
* `&`: computes bitwise AND (`ipv4`)
* `~`: computes bitwise NOT (`ipv4`)
* `|`: computes bitwise OR (`ipv4`)
* `+`: adds an offset to an ip address (`ipv4`)
* `-`: `ipv4-int`: subtracts an offset from an ip address (`ipv4`)
* `-`: `ipv4 - ipv4`: computes the difference between two ip addresses (`long`)
```
